registry/remote/repository.go

```diff
@@ -943,62 +951,153 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 	}
 }
 
-// generateDescriptor returns a descriptor generated from the response.
-func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference) (ocispec.Descriptor, error) {
+// generateBlobDescriptor returns a descriptor generated from the response.
+//
+// The following truth table aims to cover the expected GET/HEAD request outcome
+// for all possible permutations of the client/server "containing a digest".
+//
+// The client is said to "contain a digest" if the user-supplied reference string
+// is of the form that contains a digest rather than a tag.  The server, on the
+// other hand, is said to "contain a digest" if the server responded with the
+// special header `Docker-Content-Digest`.
+//
+// In this table, anything denoted with an exclamation mark indicates that the
+// true response should actually be the opposite of what's expected; for example,
+// `!PASS` means we will get a `PASS`, even though the true answer would be its
+// diametric opposite--a `FAIL`. This may seem odd, and deserves an explanation.
+// This function has blind-spots, and while it can expend power to gain sight,
+// i.e., perform the expensive validation, we chose not to.  The reason is two-
+// fold: a) we "know" that even if we say "!PASS", it will eventually fail later
+// when checks are performed, and with that assumption, we have the luxury for
+// the second point, which is b) performance.
+//
+//	 ___________________________________________________________________________________
+//	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
+//	|----+---------------+----------------+-----------------------+---------------------+
+//	| 1  | tag           | missing        | CALCULATE,PASS        | FAIL                |
+//	| 2  | tag           | presentValid   | TRUST,PASS            | TRUST,PASS          |
+//	| 3  | tag           | presentInvalid | TRUST,!PASS           | TRUST,!PASS         |
+//	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
```

Comment:
For row 4, it looks like the behavior is different for blobs and manifests. We need to keep them consistent.
When the client does `HEAD` request by digest, and the server header is missing,
1. `generateBlobDescriptor` honors the client digest
2. `generateManifestDescriptor` returns an error

In the original implementation, we honor the client digest for both blobs and manifest. See https://github.com/oras-project/oras-go/blob/main/registry/remote/repository.go#L958-L968.

Reply:
That would mean:
```
//	 ___________________________________________________________________________________
//	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
//	|----+---------------+----------------+-----------------------+---------------------+
//	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | PASS                |
 //	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
```

What would happen to row 7 then? Or like @shizhMSFT suggested earlier, I think I should drop 7-9 altogether.

Reply:
Also if they're to be in sync, then the tests should not have two variables `errExpectedOnHEAD` and `errExpectedOnGET`, rather just one `errExpected`.  Or the consistency you're referring to the case of row 4 only?

Reply:
The consistency I referred to is between blob descriptors and manifest descriptors. It makes sense that `HEAD` and `GET` behaves differently, like row 1.

Reply:
Very good point, and this is what prompted me to expand the truth table to include two new rows for Blobs.